### DISCUSSION TITLE
Issue with a short pulse during initialization of two PWM pins

### DISCUSSION THREAD

**Discussion by klinwer** [micropython/micropython#12465] [ESP32]
On a ESP32-C3-DevKitM-1 with firmware esp32c3-20230426-v1.20.0 a short pulse (~100ms) is released during the pwm initialization.

1. pwm0 = PWM(Pin(3), freq=200, duty_u16=0) without any pulse as expected.
2. pwm1 = PWM(Pin(4), freq=200, duty_u16=0) -> **A short pulse is released on Pin(3)**, Pin(4) without any pulse as expected.

I did some tests with different pins and got everytime the same result. The initialization of the second Pin releases a pulse on the fist one.

Is this caused by micropython or is it hardware related?

**Comment by robert-hh** [comment]
I could not replicate that with the nightly build. Please try the recent nightly build.

**Reply by klinwer** [reply]
I get the issue also with v1.20.0-475-g174bb28d8. How did you run your test?
I have boot.py with an import of test_2.py and in test_2.py there are these two lines in.

**Reply by klinwer** [reply]
Hi, I had a chance to measure the puls time with this script.
while True:
    pwm0 = PWM(Pin(3), freq=200, duty_u16=0)  # create PWM object from a pin
    pwm1 = PWM(Pin(4), freq=200, duty_u16=0)  # create PWM object from a pin
    pwm0.deinit()  # turn off PWM on the pin
    pwm1.deinit()
    time.sleep(0.05)

The puls time of  Pin(3) is about 0.2 ms.

**Reply by ricksorensen** [reply]
I tried this on a SEEED XIAO ESP32C3 using the ESP32_GENERIC_C3 nightly build:
`MicroPython v1.20.0-475-g174bb28d8 on 2023-09-14; ESP32C3 module with ESP32C3` 
using `GPIO3` and `GPIO4`

I did not notice the original posted problem (extra pulse when initializing second PWM), but I did notice an initial pulse that varies with frequency when creating the first PWM after a hard (power) reset.  The results (all with 0 duty cycle) for `PWM(Pin(3),freq=f,duty_u16=0)` , following a power reset were:
```table

  f         InitialPulse
  200       2.5ms     half-cycle width
  500       1  ms     half-cycle width
 1000       0.5ms     half-cycle width
 5000       0.2ms     full-cycle width
10000       0.1ms     full-cycle width
```
This initial pulse only occured with the first PWM instantiation after a power reset.  It did not occur for me after a soft (CtlD) reset.   I did try multiple pins (same for GPIO4 and GPIO5) with the same results.   I can't explain the behavior yet - maybe some timer/clock initiation and do not understand why the initial pulse width changed from half the PWM cycle width to the full width at higher frequencies.
 
Captures were driven by a rising edge on GPIO3, and the screen shots show only the leading pulse and there was only the one. 
 The logic analyzer was sampling at 10MHz for 1_000_000 samples (.1 s).   I used `mpremote` to execute the script from the command line.

f=200 capture  ` mpremote exec "from machine import PWM,Pin;p3=PWM(Pin(3),freq=200,duty_u16=0)"`
![image](https://github.com/micropython/micropython/assets/3868608/59751cc3-a2a4-4cae-a930-df753e11953c)
f=500 capture  ` mpremote exec "from machine import PWM,Pin;p3=PWM(Pin(3),freq=500,duty_u16=0)"`
![image](https://github.com/micropython/micropython/assets/3868608/7e9102d7-9815-4e4d-872d-b782910d1563)
f=1000 capture   ` mpremote exec "from machine import PWM,Pin;p3=PWM(Pin(3),freq=1000,duty_u16=0)"`
![image](https://github.com/micropython/micropython/assets/3868608/cb75b2bf-be06-4745-93fc-722703ba80b4)
f=5000 capture   ` mpremote exec "from machine import PWM,Pin;p3=PWM(Pin(3),freq=5000,duty_u16=0)"`
![image](https://github.com/micropython/micropython/assets/3868608/87e84979-c0bf-4328-b70d-59045dcd0bf7)
f=10000 capture ` mpremote exec "from machine import PWM,Pin;p3=PWM(Pin(3),freq=10000,duty_u16=0)"`
![image](https://github.com/micropython/micropython/assets/3868608/88a80188-a600-4e2e-ab10-092dbb1c400d)

**Reply by klinwer** [reply]
This confirms my observation. I have no logic analyzer to measure this in detail. Thanks for doing this.
In my application I use the pwms to dimm two different power LEDs and saw the pulse after every power on. 
The question for me is, do I really have to spend additional hardware and code to suppress the initial pulse or is there another software option?

**Reply by ricksorensen** [reply]
wink- maybe initialize a PWM with a disconnected GPIO pin first to get over the first initialization without driving a real component?

**Reply by klinwer** [reply]
That's a nice idea. My solution is to use an aditional pin:
 -  init pwm enable
    pin_pwm_enable = Pin(6), Pin.OUT)
    pin_pwm_enable.value(0)

-   Init pwm gpios
    pwm0 = PWM(Pin(3), freq=200, duty_u16=0) # create PWM object from a pin
    pwm1 = PWM(Pin(4), freq=200, duty_u16=0) # create PWM object from a pin
    pin_pwm_enable.value(1)

This is already working in combination with two AND gates.

**Reply by ricksorensen** [reply]
I saw the same results with my UM TinyS3.  I looked at the esp32 `machine_pwm.c` code.
I noticed the `configure_channel` function set the the `ledc_channel_config` with a non-zero duty cycle that looked like half.  This function is called when timers/channels are being initialzed first time during the PWM init (constructor) via `mp_machine_pwm_init_helper` function.  (line 544).  After this, the PWM freq and duty cycle are set for the particular instance.  
The constructor sequence seems to be:

- Find timer to use
- Find channel to use
- if channel not configured- `configure_channel()`
- set frequency
- if duty cycle passed in - set duty cycle to that ; else - set default duty to 50%, 10 bit resolution 

So .... I changed line 199 in `configure_channel` 
```C
        .duty =  (1 << (timers[TIMER_IDX(self->mode, self->timer)].duty_resolution)) / 2,
```
to
```C
        .duty = 0,
```
A short bout of initial testing indicates that initial pulse is absent, and if a non-zero duty cycle is given constructor the PWM starts as expected ... more evaluation is needed to look at behaviors on soft-reset, `deinit()`, and `init()`

**Reply by robert-hh** [reply]
There is a PR #10854 updating PWM for ESP32, which is present since long time. It may be interesting if that behaves identical.

**Reply by ricksorensen** [reply]
PR #10854 seems to have addressed this problem - I believe it changes the configure_channel function to set the duty to the requested duty cycle -  but I didn't see the leading pulse from the tests above.  I just used the machine_pwm.c file from that PR and not all the changes.
